aws/resource_aws_macie_s3_bucket_association.go

```diff
@@ -0,0 +1,235 @@
+		return nil
+	}
+
+	m := map[string]interface{}{}
+	if aws.StringValue(res.ClassificationType.OneTime) == macie.S3OneTimeClassificationTypeFull {
+		m["one_time"] = true
+	} else {
+		m["one_time"] = false
+	}
+	if err := d.Set("classification_type", []map[string]interface{}{m}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceAwsMacieS3BucketAssociationUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).macieconn
+
+	if d.HasChange("classification_type") {
+		req := &macie.UpdateS3ResourcesInput{
+			S3ResourcesUpdate: []*macie.S3ResourceClassificationUpdate{
+				{
+					BucketName:               aws.String(d.Get("bucket_name").(string)),
+					ClassificationTypeUpdate: &macie.ClassificationTypeUpdate{},
+				},
+			},
+		}
+		if v, ok := d.GetOk("member_account_id"); ok {
+			req.MemberAccountId = aws.String(v.(string))
+		}
+		if v, ok := d.GetOk("prefix"); ok {
+			req.S3ResourcesUpdate[0].Prefix = aws.String(v.(string))
+		}
+		req.S3ResourcesUpdate[0].ClassificationTypeUpdate.OneTime = aws.String(macieS3BucketAssociationOneTimeClassification(d))
+
+		log.Printf("[DEBUG] Updating Macie S3 bucket association: %#v", req)
+		resp, err := conn.UpdateS3Resources(req)
+		if err != nil {
+			return fmt.Errorf("Error updating Macie S3 bucket association: %s", err)
+		}
+		if len(resp.FailedS3Resources) > 0 {
+			return fmt.Errorf("Error updating Macie S3 bucket association: %s", resp.FailedS3Resources[0])
+		}
+	}
+
+	return resourceAwsMacieS3BucketAssociationRead(d, meta)
+}
+
+func resourceAwsMacieS3BucketAssociationDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).macieconn
+
+	log.Printf("[DEBUG] Deleting Macie S3 bucket association: %s", d.Id())
+
+	req := &macie.DisassociateS3ResourcesInput{
+		AssociatedS3Resources: []*macie.S3Resource{
+			{
+				BucketName: aws.String(d.Get("bucket_name").(string)),
+			},
+		},
+	}
+	if v, ok := d.GetOk("member_account_id"); ok {
+		req.MemberAccountId = aws.String(v.(string))
+	}
+	if v, ok := d.GetOk("prefix"); ok {
+		req.AssociatedS3Resources[0].Prefix = aws.String(v.(string))
+	}
+
+	resp, err := conn.DisassociateS3Resources(req)
+	if err != nil {
+		return fmt.Errorf("Error deleting Macie S3 bucket association: %s", err)
+	}
+	if len(resp.FailedS3Resources) > 0 {
+		failed := resp.FailedS3Resources[0]
+		// {
+		// 	ErrorCode: "InvalidInputException",
+		// 	ErrorMessage: "The request was rejected. The specified S3 resource (bucket or prefix) is not associated with Macie.",
+		// 	FailedItem: {
+		// 	 BucketName: "tf-macie-example-002"
+		// 	}
+		// }
+		if aws.StringValue(failed.ErrorCode) == macie.ErrCodeInvalidInputException &&
+			strings.Contains(aws.StringValue(failed.ErrorMessage), "is not associated with Macie") {
+			return nil
+		}
+		return fmt.Errorf("Error deleting Macie S3 bucket association: %s", failed)
+	}
+
+	return nil
+}
+
+func macieS3BucketAssociationId(d *schema.ResourceData) string {
```

Comment:
Nitpick: This indirection is unnecessary as its only used once for a one-liner. It seems like we should also include the member account ID so we can easily support passthrough import.